src/query/tsdb/remote/codecs.go

```diff
@@ -196,6 +202,72 @@ func encodeFetchOptions(options *storage.FetchOptions) (*rpc.FetchOptions, error
 	return result, nil
 }
 
+func encodeRestrictQueryOptionsByType(
+	o *storage.RestrictByType,
+) (*rpcpb.RestrictFetchType, error) {
+	if o == nil {
+		return nil, nil
+	}
+
+	if err := o.Validate(); err != nil {
+		return nil, err
+	}
+
+	result := &rpcpb.RestrictFetchType{}
+	switch o.MetricsType {
+	case storage.UnaggregatedMetricsType:
+		result.MetricsType = rpcpb.MetricsType_UNAGGREGATED_METRICS_TYPE
+	case storage.AggregatedMetricsType:
+		result.MetricsType = rpcpb.MetricsType_AGGREGATED_METRICS_TYPE
+
+		storagePolicyProto, err := o.StoragePolicy.Proto()
+		if err != nil {
+			return nil, err
+		}
+
+		result.MetricsStoragePolicy = storagePolicyProto
+	}
+
+	return result, nil
+}
+
+func encodeRestrictQueryOptionsByTag(
+	o *storage.RestrictByTag,
+) (*rpcpb.RestrictFetchTags, error) {
+	if o == nil {
+		return nil, nil
+	}
+
+	matchers, err := encodeTagMatchers(o.GetMatchers())
+	if err != nil {
+		return nil, err
+	}
+
+	return &rpcpb.RestrictFetchTags{
+		Restrict: matchers,
+		Strip:    o.Strip,
+	}, nil
+}
+
+func encodeRestrictQueryOptions(
+	o *storage.RestrictQueryOptions,
+) (*rpcpb.RestrictQueryOptions, error) {
```

Comment:
do we need to check `o == nil` here as well?

Reply:
Technically it will create a RestrictQueryOptions with the internals being nil, but having the RestrictQueryOptions probably makes more sense :+1: